File: apps/mobile/GuiltyPartyCompanion/NetworkClient.swift
import Foundation

enum CompanionEnvironment {
    static let joinURL = URL(string: "https://api.test.guiltyparty.app/api/v1/join")!
    static let webSocketURL = URL(string: "wss://api.test.guiltyparty.app/ws/v1")!
    static let controlSubprotocol = "guiltyparty.control.v1"
    static let applicationID = "companion_ios"
    static let applicationVersion = "0.1.0"
    static let buildNumber: Int64 = 1
    static let maximumResponseBytes = 262_144

    static func shortRequestConfiguration() -> URLSessionConfiguration {
        let configuration = baseEphemeralConfiguration()
        configuration.timeoutIntervalForRequest = 15
        configuration.timeoutIntervalForResource = 20
        return configuration
    }

    static func webSocketConfiguration() -> URLSessionConfiguration {
        baseEphemeralConfiguration()
    }

    private static func baseEphemeralConfiguration() -> URLSessionConfiguration {
        let configuration = URLSessionConfiguration.ephemeral
        configuration.urlCache = nil
        configuration.requestCachePolicy = .reloadIgnoringLocalAndRemoteCacheData
        configuration.httpShouldSetCookies = false
        configuration.httpCookieAcceptPolicy = .never
        return configuration
    }
}

final class RejectingRedirectDelegate: NSObject, URLSessionTaskDelegate, @unchecked Sendable {
    func urlSession(
        _ session: URLSession,
        task: URLSessionTask,
        willPerformHTTPRedirection response: HTTPURLResponse,
        newRequest request: URLRequest,
        completionHandler: @escaping (URLRequest?) -> Void
    ) {
        completionHandler(nil)
    }
}

enum JoinClient {
    static func request(invitation: Invitation, displayName: String) throws -> URLRequest {
        let nickname = displayName.trimmingCharacters(in: .whitespacesAndNewlines)
        guard (1...80).contains(nickname.unicodeScalars.count),
              nickname.range(of: "^[^\\u0000-\\u001F\\u007F-\\u009F]+$", options: .regularExpression) != nil
        else {
            throw SessionModelError.invalidDisplayName
        }

        let clientBuild = try GPV1ClientBuild(
            applicationId: GPV1FeatureIdentifier(CompanionEnvironment.applicationID),
            applicationVersion: CompanionEnvironment.applicationVersion,
            buildNumber: CompanionEnvironment.buildNumber
        )
        let endpoint = try GPV1EndpointRegistration(
            capabilities: [
                GPV1FeatureIdentifier("private_display"),
                GPV1FeatureIdentifier("touch_input")
            ],
            clientBuild: clientBuild,
            platform: GPV1FeatureIdentifier("ios_companion")
        )
        let body = GPV1JoinRequest.participant(
            try GPV1JoinRequestParticipant(
                displayName: nickname,
                endpoint: endpoint,
                kind: "participant",
                protocolVersion: GPV1ProtocolVersion()
            )
        )

        var request = URLRequest(
            url: CompanionEnvironment.joinURL,
            cachePolicy: .reloadIgnoringLocalAndRemoteCacheData,
            timeoutInterval: 15
        )
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(invitation.sessionID, forHTTPHeaderField: "X-GP-Session-ID")
        request.setValue("Pairing \(invitation.pairingProof)", forHTTPHeaderField: "Authorization")
        request.setValue("no-store", forHTTPHeaderField: "Cache-Control")
        request.httpBody = try JSONEncoder().encode(body)
        return request
    }

    static func join(invitation: Invitation, displayName: String) async throws -> SessionAuthority {
        let request = try request(invitation: invitation, displayName: displayName)
        let session = URLSession(
            configuration: CompanionEnvironment.shortRequestConfiguration(),
            delegate: RejectingRedirectDelegate(),
            delegateQueue: nil
        )
        defer { session.invalidateAndCancel() }
        let (data, response) = try await session.data(for: request)
        guard data.count <= CompanionEnvironment.maximumResponseBytes,
              let http = response as? HTTPURLResponse
        else {
            throw SessionModelError.invalidResponse
        }
        guard http.statusCode == 200 else {
            throw classifyHTTPFailure(status: http.statusCode, data: data)
        }
        guard http.value(forHTTPHeaderField: "Cache-Control")?.lowercased().contains("no-store") == true,
              case .authorizationHeader(let joined) = try JSONDecoder().decode(
                GPV1RemoteFriendsJoinResponse.self,
                from: data
              ),
              joined.sessionId.value == invitation.sessionID,
              let participantID = joined.participantId?.value,
              joined.authorityExpiresAtUnixMs > Int64(Date().timeIntervalSince1970 * 1_000)
        else {
            throw SessionModelError.invalidResponse
        }

        return SessionAuthority(
            sessionID: joined.sessionId.value,
            endpointID: joined.endpointId.value,
            participantID: participantID,
            bearer: joined.token.value,
            expiresAtUnixMilliseconds: joined.authorityExpiresAtUnixMs,
            primaryAuthorityGeneration: joined.primaryAuthorityGeneration,
            gameplayLanguage: invitation.gameplayLanguage
        )
    }

    private static func classifyHTTPFailure(status: Int, data: Data) -> SessionModelError {
        let problem = try? JSONDecoder().decode(GPV1ProblemDetails.self, from: data)
        let code = problem?.code ?? ""
        if status == 410 || code.contains("expired") || code.contains("revoked") || code.contains("invitation") {
            return .expiredOrRevoked
        }
        if code == "unsupported_client_build" || code == "upgrade_required" {
            return .commandRejected("This development build is no longer supported by the test service.")
        }
        return .commandRejected(problem?.title ?? "The session could not be joined.")
    }
}

enum WebSocketRequestBuilder {
    static func request(authority: SessionAuthority) -> URLRequest {
        var request = URLRequest(
            url: CompanionEnvironment.webSocketURL,
            cachePolicy: .reloadIgnoringLocalAndRemoteCacheData,
            timeoutInterval: 10
        )
        request.setValue("Bearer \(authority.bearer)", forHTTPHeaderField: "Authorization")
        request.setValue(
            CompanionEnvironment.controlSubprotocol,
            forHTTPHeaderField: "Sec-WebSocket-Protocol"
        )
        request.setValue("no-store", forHTTPHeaderField: "Cache-Control")
        return request
    }
}

enum SocketLifecycleEvent: Equatable, Sendable {
    case opened(subprotocol: String?)
    case closed(code: Int)
    case failed
}

final class WebSocketDelegate: NSObject, URLSessionWebSocketDelegate, @unchecked Sendable {
    private let handler: @Sendable (SocketLifecycleEvent) -> Void

    init(handler: @escaping @Sendable (SocketLifecycleEvent) -> Void) {
        self.handler = handler
    }

    func urlSession(
        _ session: URLSession,
        webSocketTask: URLSessionWebSocketTask,
        didOpenWithProtocol protocol: String?
    ) {
        handler(.opened(subprotocol: `protocol`))
    }

    func urlSession(
        _ session: URLSession,
        webSocketTask: URLSessionWebSocketTask,
        didCloseWith closeCode: URLSessionWebSocketTask.CloseCode,
        reason: Data?
    ) {
        handler(.closed(code: closeCode.rawValue))
    }

    func urlSession(
        _ session: URLSession,
        task: URLSessionTask,
        didCompleteWithError error: (any Error)?
    ) {
        if error != nil {
            handler(.failed)
        }
    }

    func urlSession(
        _ session: URLSession,
        task: URLSessionTask,
        willPerformHTTPRedirection response: HTTPURLResponse,
        newRequest request: URLRequest,
        completionHandler: @escaping (URLRequest?) -> Void
    ) {
        completionHandler(nil)
        handler(.failed)
    }
}

final class FirstPartyWebSocket: @unchecked Sendable {
    private let delegate: WebSocketDelegate
    private let session: URLSession
    private let task: URLSessionWebSocketTask

    init(
        request: URLRequest,
        lifecycleHandler: @escaping @Sendable (SocketLifecycleEvent) -> Void
    ) {
        let delegate = WebSocketDelegate(handler: lifecycleHandler)
        let session = URLSession(
            configuration: CompanionEnvironment.webSocketConfiguration(),
            delegate: delegate,
            delegateQueue: nil
        )
        self.delegate = delegate
        self.session = session
        task = session.webSocketTask(with: request)
        task.maximumMessageSize = CompanionEnvironment.maximumResponseBytes
    }

    var closeCode: Int {
        task.closeCode.rawValue
    }

    func start() {
        task.resume()
    }

    func send(_ data: Data) async throws {
        try await task.send(WebSocketControlFrame.message(from: data))
    }

    func receive() async throws -> Data {
        switch try await task.receive() {
        case .data(let data):
            guard data.count <= CompanionEnvironment.maximumResponseBytes else {
                throw SessionModelError.protocolViolation
            }
            return data
        case .string(let text):
            guard let data = text.data(using: .utf8),
                  data.count <= CompanionEnvironment.maximumResponseBytes
            else {
                throw SessionModelError.protocolViolation
            }
            return data
        @unknown default:
            throw SessionModelError.protocolViolation
        }
    }

    func ping() async throws {
        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
            task.sendPing { error in
                if let error {
                    continuation.resume(throwing: error)
                } else {
                    continuation.resume()
                }
            }
        }
    }

    func cancel() {
        task.cancel(with: .goingAway, reason: nil)
        session.invalidateAndCancel()
    }
}

enum WebSocketControlFrame {
    static func message(from data: Data) throws -> URLSessionWebSocketTask.Message {
        guard let text = String(data: data, encoding: .utf8) else {
            throw SessionModelError.protocolViolation
        }
        return .string(text)
    }
}

enum IncomingControlEvent: Equatable, Sendable {
    case projection(GPV1ProjectionEnvelope)
    case commandResult(GPV1CommandResultEnvelope)
    case error(GPV1ErrorEnvelope)
}

enum ControlPlaneCodec {
    static func decodeServerEvent(
        _ data: Data,
        authority: SessionAuthority
    ) throws -> IncomingControlEvent {
        let envelope = try JSONDecoder().decode(GPV1ServerEnvelope.self, from: data)
        switch envelope {
        case .projection(let projection):
            try validateContext(
                sessionID: projection.sessionId.value,
                endpointID: projection.endpointId.value,
                authority: authority
            )
            return .projection(projection)
        case .commandResult(let result):
            try validateContext(
                sessionID: result.sessionId.value,
                endpointID: result.endpointId.value,
                authority: authority
            )
            return .commandResult(result)
        case .error(let error):
            guard let sessionID = error.sessionId?.value,
                  let endpointID = error.endpointId?.value
            else {
                throw SessionModelError.protocolViolation
            }
            try validateContext(
                sessionID: sessionID,
                endpointID: endpointID,
                authority: authority
            )
            return .error(error)
        case .aiSuggestion:
            throw SessionModelError.protocolViolation
        }
    }

    static func projectionRequest(authority: SessionAuthority, messageID: String) throws -> Data {
        let envelope = GPV1ClientEnvelope.getProjection(
            try GPV1GetProjectionEnvelope(
                endpointId: GPV1Identifier(authority.endpointID),
                messageId: GPV1MessageIdentifier(messageID),
                payload: GPV1GetProjectionEnvelopePayload(),
                protocolVersion: GPV1ProtocolVersion(),
                sessionId: GPV1Identifier(authority.sessionID),
                type: "get_projection"
            )
        )
        return try JSONEncoder().encode(envelope)
    }

    private static func validateContext(
        sessionID: String,
        endpointID: String,
        authority: SessionAuthority
    ) throws {
        guard sessionID == authority.sessionID, endpointID == authority.endpointID else {
            throw SessionModelError.protocolViolation
        }
    }
}

struct OutboundCommand: Equatable, Sendable {
    let messageID: String
    let idempotencyID: String
    let targetCharacterID: String
}

struct CommandFactory: Sendable {
    private let makeUUID: @Sendable () -> UUID

    init(makeUUID: @escaping @Sendable () -> UUID = UUID.init) {
        self.makeUUID = makeUUID
    }

    func castVote(targetCharacterID: String) throws -> OutboundCommand {
        let target = targetCharacterID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard (1...128).contains(target.unicodeScalars.count) else {
            throw SessionModelError.commandRejected("Enter the target character identifier supplied for this test.")
        }
        return OutboundCommand(
            messageID: identifier(prefix: "msg"),
            idempotencyID: identifier(prefix: "cmd"),
            targetCharacterID: target
        )
    }

    func retry(_ command: OutboundCommand) -> OutboundCommand {
        OutboundCommand(
            messageID: identifier(prefix: "msg"),
            idempotencyID: command.idempotencyID,
            targetCharacterID: command.targetCharacterID
        )
    }

    func encode(_ command: OutboundCommand, authority: SessionAuthority) throws -> Data {
        let vote = try GPV1CastVoteCommand(
            targetCharacterId: GPV1Identifier(command.targetCharacterID),
            type: "cast_vote"
        )
        let envelope = GPV1ClientEnvelope.submitCommand(
            try GPV1SubmitCommandEnvelope(
                endpointId: GPV1Identifier(authority.endpointID),
                idempotencyId: GPV1MessageIdentifier(command.idempotencyID),
                messageId: GPV1MessageIdentifier(command.messageID),
                payload: GPV1SubmitCommandEnvelopePayload(command: .castVote(vote)),
                primaryAuthorityGeneration: authority.primaryAuthorityGeneration,
                protocolVersion: GPV1ProtocolVersion(),
                sessionId: GPV1Identifier(authority.sessionID),
                type: "submit_command"
            )
        )
        return try JSONEncoder().encode(envelope)
    }

    private func identifier(prefix: String) -> String {
        "\(prefix)_\(makeUUID().uuidString.replacingOccurrences(of: "-", with: "").lowercased())"
    }
}

File: apps/mobile/android/app/src/main/java/com/guiltyparty/companion/NetworkClient.kt
package com.guiltyparty.companion

import guiltyparty.contracts.v1.GPV1ClientBuild
import guiltyparty.contracts.v1.GPV1EndpointRegistration
import guiltyparty.contracts.v1.GPV1FeatureIdentifier
import guiltyparty.contracts.v1.GPV1Identifier
import guiltyparty.contracts.v1.GPV1JoinRequest
import guiltyparty.contracts.v1.GPV1JoinRequestParticipant
import guiltyparty.contracts.v1.GPV1MessageIdentifier
import guiltyparty.contracts.v1.GPV1ParticipantResumeRequest
import guiltyparty.contracts.v1.GPV1ProblemDetails
import guiltyparty.contracts.v1.GPV1ProtocolVersion
import guiltyparty.contracts.v1.GPV1RemoteFriendsJoinResponse
import guiltyparty.contracts.v1.GPV1RemoteNativeResumeResponse
import guiltyparty.contracts.v1.GPV1ResumedCommandStatus
import okhttp3.Call
import okhttp3.Callback
import okhttp3.CookieJar
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString
import java.io.IOException
import java.util.concurrent.TimeUnit

interface AdmissionCallback {
    fun onSuccess(admission: ParticipantAdmission)
    fun onFailure(failure: CompanionFailure)
}

interface SocketCallback {
    fun onOpen(socketId: String)
    fun onText(socketId: String, text: String)
    fun onClosed(socketId: String, code: Int)
    fun onFailure(socketId: String, failure: SocketFailure)
}

enum class SocketFailure { AMBIGUOUS_TRANSPORT, PROTOCOL_VIOLATION }

class NetworkClient {
    internal val httpClient = baseBuilder()
        .callTimeout(20, TimeUnit.SECONDS)
        .connectTimeout(10, TimeUnit.SECONDS)
        .readTimeout(15, TimeUnit.SECONDS)
        .writeTimeout(15, TimeUnit.SECONDS)
        .build()
    internal val socketClient = baseBuilder()
        .connectTimeout(10, TimeUnit.SECONDS)
        .pingInterval(15, TimeUnit.SECONDS)
        .readTimeout(0, TimeUnit.MILLISECONDS)
        .build()

    fun join(
        endpoint: ServerEndpoint,
        invitation: Invitation,
        displayName: String,
        callback: AdmissionCallback,
    ): Call {
        val request = buildJoinRequest(endpoint, invitation, displayName)
        val call = httpClient.newCall(request)
        call.enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                callback.onFailure(CompanionFailure(FailureKind.CONNECTION_UNAVAILABLE))
            }

            override fun onResponse(call: Call, response: Response) {
                response.use {
                    try {
                        val text = responseText(it)
                        if (!it.isSuccessful) throw classifyHttpFailure(it.code, text)
                        requireNoStore(it)
                        callback.onSuccess(decodeJoinResponse(text, invitation, endpoint))
                    } catch (failure: CompanionFailure) {
                        callback.onFailure(failure)
                    } catch (_: Exception) {
                        callback.onFailure(CompanionFailure(FailureKind.INVALID_RESPONSE))
                    }
                }
            }
        })
        return call
    }

    internal fun buildJoinRequest(
        endpoint: ServerEndpoint,
        invitation: Invitation,
        displayName: String,
    ): Request {
        val nickname = displayName.trim()
        if (nickname.codePointCount(0, nickname.length) !in 1..80 ||
            nickname.any { it.code in 0x00..0x1f || it.code in 0x7f..0x9f }
        ) {
            throw CompanionFailure(FailureKind.INVALID_DISPLAY_NAME)
        }
        val body = GPV1JoinRequest.Participant(
            GPV1JoinRequestParticipant(
                displayName = nickname,
                endpoint = GPV1EndpointRegistration(
                    capabilities = listOf(
                        GPV1FeatureIdentifier("private_display"),
                        GPV1FeatureIdentifier("touch_input"),
                    ),
                    clientBuild = clientBuild(),
                    platform = GPV1FeatureIdentifier("android_companion"),
                ),
                kind = "participant",
                protocolVersion = GPV1ProtocolVersion("1.0"),
            ),
        )
        return Request.Builder()
            .url(endpoint.joinUrl)
            .header("Accept", JSON_MEDIA_TYPE)
            .header("Cache-Control", "no-store")
            .header("X-GP-Session-ID", invitation.sessionId)
            .header("Authorization", "Pairing ${invitation.pairingProof}")
            .post(ContractJson.stringify(body.toJson()).toRequestBody(JSON_TYPE))
            .build()
    }

    internal fun decodeJoinResponse(
        text: String,
        invitation: Invitation,
        endpoint: ServerEndpoint,
        nowUnixMs: Long = System.currentTimeMillis(),
    ): ParticipantAdmission {
        val result = GPV1RemoteFriendsJoinResponse.from(ContractJson.parse(text))
        val joined = (result as? GPV1RemoteFriendsJoinResponse.AuthorizationHeader)?.value
            ?: throw CompanionFailure(FailureKind.INVALID_RESPONSE)
        val participant = joined.participantId?.value
            ?: throw CompanionFailure(FailureKind.INVALID_RESPONSE)
        if (joined.sessionId.value != invitation.sessionId ||
            joined.authorityExpiresAtUnixMs <= nowUnixMs ||
            joined.resumeExpiresAtUnixMs <= nowUnixMs
        ) {
            throw CompanionFailure(FailureKind.INVALID_RESPONSE)
        }
        return ParticipantAdmission(
            authority = SessionAuthority(
                joined.sessionId.value,
                joined.endpointId.value,
                participant,
                joined.token.value,
                joined.authorityExpiresAtUnixMs,
                joined.primaryAuthorityGeneration,
                invitation.gameplayLanguage,
                endpoint,
            ),
            resumeCredential = StoredResumeCredential(
                joined.resumeToken.value,
                null,
                joined.sessionId.value,
                joined.endpointId.value,
                participant,
                joined.resumeExpiresAtUnixMs,
                joined.primaryAuthorityGeneration,
                joined.serverSequence,
                nowUnixMs,
                emptyList(),
                invitation.gameplayLanguage,
                endpoint.origin,
            ),
        )
    }

    fun resume(credential: StoredResumeCredential, callback: AdmissionCallback): Call {
        val request = buildResumeRequest(credential)
        val replacement = credential.pendingReplacementToken
            ?: throw CompanionFailure(FailureKind.PROTOCOL_VIOLATION)
        val endpoint = ServerEndpoint.parse(credential.serverOrigin, BuildConfig.DEBUG)
        val call = httpClient.newCall(request)
        call.enqueue(object : Callback {
            override fun onFailure(call: Call, e: IOException) {
                callback.onFailure(CompanionFailure(FailureKind.CONNECTION_UNAVAILABLE))
            }

            override fun onResponse(call: Call, response: Response) {
                response.use {
                    try {
                        val text = responseText(it)
                        if (!it.isSuccessful) throw classifyHttpFailure(it.code, text)
                        requireNoStore(it)
                        val resumed = GPV1RemoteNativeResumeResponse.from(ContractJson.parse(text))
                        validateResume(resumed, credential, replacement)
                        callback.onSuccess(admissionFromResume(resumed, credential, endpoint))
                    } catch (failure: CompanionFailure) {
                        callback.onFailure(failure)
                    } catch (_: Exception) {
                        callback.onFailure(CompanionFailure(FailureKind.INVALID_RESPONSE))
                    }
                }
            }
        })
        return call
    }

    internal fun buildResumeRequest(credential: StoredResumeCredential): Request {
        val replacement = credential.pendingReplacementToken
        if (replacement.isNullOrBlank() || replacement == credential.token) {
            throw CompanionFailure(FailureKind.PROTOCOL_VIOLATION)
        }
        val endpoint = ServerEndpoint.parse(credential.serverOrigin, BuildConfig.DEBUG)
        val body = GPV1ParticipantResumeRequest(
            clientBuild = clientBuild(),
            endpointId = GPV1Identifier(credential.endpointId),
            lastServerSequence = credential.lastServerSequence,
            participantId = GPV1Identifier(credential.participantId),
            pendingIdempotencyIds = credential.pendingIdempotencyIds.map(::GPV1MessageIdentifier),
            primaryAuthorityGeneration = credential.primaryAuthorityGeneration,
            protocolVersion = GPV1ProtocolVersion("1.0"),
            sessionId = GPV1Identifier(credential.sessionId),
        )
        return Request.Builder()
            .url(endpoint.resumeUrl)
            .header("Accept", JSON_MEDIA_TYPE)
            .header("Cache-Control", "no-store")
            .header("Authorization", "Resume ${credential.token}")
            .header("X-GP-Replacement-Resume", replacement)
            .post(ContractJson.stringify(body.toJson()).toRequestBody(JSON_TYPE))
            .build()
    }

    internal fun decodeResumeResponse(
        text: String,
        previous: StoredResumeCredential,
        nowUnixMs: Long = System.currentTimeMillis(),
    ): ParticipantAdmission {
        val replacement = previous.pendingReplacementToken
            ?: throw CompanionFailure(FailureKind.PROTOCOL_VIOLATION)
        val resumed = GPV1RemoteNativeResumeResponse.from(ContractJson.parse(text))
        validateResume(resumed, previous, replacement, nowUnixMs)
        return admissionFromResume(
            resumed,
            previous,
            ServerEndpoint.parse(previous.serverOrigin, BuildConfig.DEBUG),
            nowUnixMs,
        )
    }

    fun openSocket(authority: SessionAuthority, socketId: String, callback: SocketCallback): WebSocket {
        val request = Request.Builder()
            .url(authority.serverEndpoint.webSocketUrl)
            .header("Authorization", "Bearer ${authority.bearer}")
            .header("Sec-WebSocket-Protocol", CONTROL_SUBPROTOCOL)
            .header("Cache-Control", "no-store")
            .build()
        return socketClient.newWebSocket(request, object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                if (response.header("Sec-WebSocket-Protocol") != CONTROL_SUBPROTOCOL) {
                    webSocket.close(1002, null)
                    callback.onFailure(socketId, SocketFailure.PROTOCOL_VIOLATION)
                    return
                }
                callback.onOpen(socketId)
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                if (text.toByteArray(Charsets.UTF_8).size > MAXIMUM_RESPONSE_BYTES) {
                    webSocket.close(1009, null)
                    callback.onFailure(socketId, SocketFailure.PROTOCOL_VIOLATION)
                } else {
                    callback.onText(socketId, text)
                }
            }

            override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
                webSocket.close(1003, null)
                callback.onFailure(socketId, SocketFailure.PROTOCOL_VIOLATION)
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                callback.onClosed(socketId, code)
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                callback.onFailure(socketId, SocketFailure.AMBIGUOUS_TRANSPORT)
            }
        })
    }

    private fun admissionFromResume(
        resumed: GPV1RemoteNativeResumeResponse,
        previous: StoredResumeCredential,
        endpoint: ServerEndpoint,
        nowUnixMs: Long = System.currentTimeMillis(),
    ) = ParticipantAdmission(
        authority = SessionAuthority(
            resumed.sessionId.value,
            resumed.endpointId.value,
            resumed.participantId.value,
            resumed.token.value,
            resumed.authorityExpiresAtUnixMs,
            resumed.primaryAuthorityGeneration,
            previous.gameplayLanguage,
            endpoint,
        ),
        resumeCredential = StoredResumeCredential(
            resumed.resumeToken.value,
            null,
            resumed.sessionId.value,
            resumed.endpointId.value,
            resumed.participantId.value,
            resumed.resumeExpiresAtUnixMs,
            resumed.primaryAuthorityGeneration,
            resumed.serverSequence,
            nowUnixMs,
            emptyList(),
            previous.gameplayLanguage,
            endpoint.origin,
        ),
    )

    private fun validateResume(
        resumed: GPV1RemoteNativeResumeResponse,
        previous: StoredResumeCredential,
        replacement: String,
        nowUnixMs: Long = System.currentTimeMillis(),
    ) {
        val requested = previous.pendingIdempotencyIds.toSet()
        val results = resumed.pendingCommandResults.associate { result ->
            when (result) {
                is GPV1ResumedCommandStatus.Accepted -> result.value.idempotencyId.value to result.value.serverSequence
                is GPV1ResumedCommandStatus.Rejected -> result.value.idempotencyId.value to result.value.serverSequence
                is GPV1ResumedCommandStatus.Unknown -> result.value.idempotencyId.value to null
            }
        }
        if (resumed.sessionId.value != previous.sessionId ||
            resumed.endpointId.value != previous.endpointId ||
            resumed.participantId.value != previous.participantId ||
            resumed.resumeToken.value != replacement ||
            resumed.primaryAuthorityGeneration <= previous.primaryAuthorityGeneration ||
            resumed.authorityExpiresAtUnixMs <= nowUnixMs ||
            resumed.resumeExpiresAtUnixMs <= nowUnixMs ||
            resumed.serverSequence < previous.lastServerSequence ||
            results.keys != requested ||
            results.values.filterNotNull().any { it > resumed.serverSequence }
        ) {
            throw CompanionFailure(FailureKind.INVALID_RESPONSE)
        }
    }

    private fun responseText(response: Response): String {
        val body = response.body ?: throw CompanionFailure(FailureKind.INVALID_RESPONSE)
        if (body.contentLength() > MAXIMUM_RESPONSE_BYTES) {
            throw CompanionFailure(FailureKind.INVALID_RESPONSE)
        }
        val bytes = body.byteStream().readNBytes(MAXIMUM_RESPONSE_BYTES + 1)
        if (bytes.size > MAXIMUM_RESPONSE_BYTES) {
            throw CompanionFailure(FailureKind.INVALID_RESPONSE)
        }
        return ContractJson.decodeUtf8(bytes)
    }

    private fun requireNoStore(response: Response) {
        if (!response.header("Cache-Control").orEmpty().lowercase().contains("no-store")) {
            throw CompanionFailure(FailureKind.INVALID_RESPONSE)
        }
    }

    private fun classifyHttpFailure(status: Int, text: String): CompanionFailure {
        val problem = runCatching { GPV1ProblemDetails.from(ContractJson.parse(text)) }.getOrNull()
        val code = problem?.code.orEmpty()
        return when {
            code == "session_ended" -> CompanionFailure(FailureKind.SESSION_ENDED)
            status == 410 || status == 401 || code.contains("expired") ||
                code.contains("revoked") || code.contains("invitation") || code.contains("resume") ->
                CompanionFailure(FailureKind.EXPIRED_OR_REVOKED)
            status == 429 || status >= 500 -> CompanionFailure(FailureKind.CONNECTION_UNAVAILABLE)
            code == "unsupported_client_build" || code == "upgrade_required" ->
                CompanionFailure(
                    FailureKind.COMMAND_REJECTED,
                    "This development build is no longer supported by the test service.",
                )
            else -> CompanionFailure(
                FailureKind.COMMAND_REJECTED,
                problem?.title ?: "The session could not be joined.",
            )
        }
    }

    private fun clientBuild() = GPV1ClientBuild(
        applicationId = GPV1FeatureIdentifier("companion_android"),
        applicationVersion = BuildConfig.VERSION_NAME.substringBefore('-'),
        buildNumber = BuildConfig.VERSION_CODE.toLong(),
    )

    private fun baseBuilder() = OkHttpClient.Builder()
        .cache(null)
        .cookieJar(CookieJar.NO_COOKIES)
        .followRedirects(false)
        .followSslRedirects(false)
        .retryOnConnectionFailure(false)

    companion object {
        const val CONTROL_SUBPROTOCOL = "guiltyparty.control.v1"
        const val MAXIMUM_RESPONSE_BYTES = 262_144
        private const val JSON_MEDIA_TYPE = "application/json"
        private val JSON_TYPE = JSON_MEDIA_TYPE.toMediaType()
    }
}
